auth/device_flow.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2022 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+// DeviceCode holds information about the authorization-in-progress.
+type DeviceCode struct {
+	UserCode        string `json:"user_code"`        // UserCode is the code presented to users
+	VerificationURI string `json:"verification_uri"` // VerificationURI is the URI where users will need to confirm the code
+	DeviceCode      string `json:"device_code"`      // DeviceCode is the internal code to confirm the status of the flow
+	ExpiresIn       int    `json:"expires_in"`       // ExpiresIn when the code will expire
+	Interval        int    `json:"interval"`         // Interval who often to verify the status of the code
+
+	timeNow   func() time.Time
+	timeSleep func(time.Duration)
+}
+
+// RequestCode initiates the authorization flow by requesting a code.
+func (c Config) RequestCode(ctx context.Context) (*DeviceCode, *atlas.Response, error) {
+	req, err := c.NewRequest(ctx, http.MethodPost, "api/private/unauth/account/device/authorize",
+		url.Values{
+			"client_id": {c.ClientID},
+			"scope":     {strings.Join(c.Scopes, " ")},
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	var r *DeviceCode
+	resp, err2 := c.Do(ctx, req, &r)
+	return r, resp, err2
+}
+
+// ErrTimeout is thrown when polling the server for the granted token has timed out.
+var ErrTimeout = errors.New("authentication timed out")
+
+// PollToken polls the server until an access token is granted or denied.
+func (c Config) PollToken(ctx context.Context, code *DeviceCode) (*Token, *atlas.Response, error) {
+	timeNow := code.timeNow
+	if timeNow == nil {
+		timeNow = time.Now
+	}
```

Comment:
[question for my own learning] will this ever trigger? `timeNow func() time.Time` will always return time?

Reply:
this is a reference to the function `time.Now`, line 93 then invokes the function which will return "now" at the moment of invocation, this patters is so now can be mocked by passing a `code` with a different "now"